File: src/utils/template.py
import json
import sys
from typing import List, Optional, Dict, Any, Union
from collections import defaultdict, deque

# Definition for singly-linked list
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# Definition for binary tree
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class IO:
    """
    Input/Output utility class for various data types
    Handles reading from stdin and writing to stdout for different types
    """
    
    @staticmethod
    def read_bool():
        return input().strip().lower() in ['true', '1']
    
    @staticmethod
    def read_int():
        return int(input().strip())
    
    @staticmethod
    def read_float():
        return float(input().strip())
    
    @staticmethod
    def read_string():
        return input().strip()
    
    @staticmethod
    def read_list_int():
        line = input().strip()
        if not line or line == "[]":
            return []
        return list(map(int, line.strip('[]').split(',')))
    
    @staticmethod
    def read_list_string():
        line = input().strip()
        if not line or line == "[]":
            return []
        # Handle quoted strings in list
        line = line.strip('[]')
        if not line:
            return []
        items = []
        current = ""
        in_quotes = False
        for char in line:
            if char == '"' and not in_quotes:
                in_quotes = True
            elif char == '"' and in_quotes:
                in_quotes = False
                items.append(current)
                current = ""
            elif char == ',' and not in_quotes:
                if current and current != ' ':
                    items.append(current.strip())
                current = ""
            else:
                if not (char == ' ' and not in_quotes and not current):
                    current += char
        if current.strip():
            items.append(current.strip())
        return items
    
    @staticmethod
    def read_list_list_int():
        line = input().strip()
        if not line or line == "[]":
            return []
        # Parse nested list format like [[1,2],[3,4]]
        result = []
        line = line.strip('[]')
        if not line:
            return []
        
        i = 0
        while i < len(line):
            if line[i] == '[':
                j = i + 1
                bracket_count = 1
                while j < len(line) and bracket_count > 0:
                    if line[j] == '[':
                        bracket_count += 1
                    elif line[j] == ']':
                        bracket_count -= 1
                    j += 1
                sublist_str = line[i+1:j-1]
                if sublist_str:
                    result.append(list(map(int, sublist_str.split(','))))
                else:
                    result.append([])
                i = j
                while i < len(line) and line[i] in ', ':
                    i += 1
            else:
                i += 1
        return result
    
    @staticmethod
    def read_linked_list():
        values = IO.read_list_int()
        if not values:
            return None
        
        head = ListNode(values[0])
        current = head
        for val in values[1:]:
            current.next = ListNode(val)
            current = current.next
        return head
    
    @staticmethod
    def read_binary_tree():
        line = input().strip()
        if not line or line == "[]" or line == "null":
            return None
        
        # Parse tree in level-order format like [3,9,20,null,null,15,7]
        line = line.strip('[]')
        if not line:
            return None
        
        values = []
        current = ""
        for char in line:
            if char == ',':
                val = current.strip()
                values.append(None if val == "null" else int(val))
                current = ""
            else:
                current += char
        if current.strip():
            val = current.strip()
            values.append(None if val == "null" else int(val))
        
        if not values or values[0] is None:
            return None
        
        root = TreeNode(values[0])
        queue = [root]
        i = 1
        
        while queue and i < len(values):
            node = queue.pop(0)
            
            # Left child
            if i < len(values):
                if values[i] is not None:
                    node.left = TreeNode(values[i])
                    queue.append(node.left)
                i += 1
            
            # Right child
            if i < len(values):
                if values[i] is not None:
                    node.right = TreeNode(values[i])
                    queue.append(node.right)
                i += 1
        
        return root
    
    @staticmethod
    def write_bool(x):
        print("True" if x else "False", end="")
    
    @staticmethod
    def write_int(x):
        print(x, end="")
    
    @staticmethod
    def write_float(x):
        print(x, end="")
    
    @staticmethod
    def write_string(x):
        print(x, end="")
    
    @staticmethod
    def write_list(x):
        if isinstance(x[0], list):
            # Handle nested list
            print("[", end="")
            for i, sublist in enumerate(x):
                if i > 0:
                    print(",", end="")
                IO.write_list(sublist)
            print("]", end="")
        else:
            print("[", end="")
            for i, item in enumerate(x):
                if i > 0:
                    print(",", end="")
                if isinstance(item, str):
                    print(f'"{item}"', end="")
                else:
                    print(item, end="")
            print("]", end="")
    
    @staticmethod
    def write_linked_list(head):
        values = []
        current = head
        while current:
            values.append(current.val)
            current = current.next
        IO.write_list(values)
    
    @staticmethod
    def write_binary_tree(root):
        if not root:
            print("[]", end="")
            return
        
        result = []
        queue = [root]
        
        while queue:
            node = queue.pop(0)
            if node:
                result.append(node.val)
                queue.append(node.left)
                queue.append(node.right)
            else:
                result.append(None)
        
        # Remove trailing nulls
        while result and result[-1] is None:
            result.pop()
        
        print("[", end="")
        for i, val in enumerate(result):
            if i > 0:
                print(",", end="")
            if val is None:
                print("null", end="")
            else:
                print(val, end="")
        print("]", end="")

# Convenience functions
def input_data(data_type):
    if data_type == "int":
        return IO.read_int()
    elif data_type == "float":
        return IO.read_float()
    elif data_type == "string":
        return IO.read_string()
    elif data_type == "bool":
        return IO.read_bool()
    elif data_type == "List[int]":
        return IO.read_list_int()
    elif data_type == "List[str]":
        return IO.read_list_string()
    elif data_type == "List[List[int]]":
        return IO.read_list_list_int()
    elif data_type == "ListNode":
        return IO.read_linked_list()
    elif data_type == "TreeNode":
        return IO.read_binary_tree()
    else:
        return input().strip()

def output_data(x):
    if isinstance(x, bool):
        IO.write_bool(x)
    elif isinstance(x, int):
        IO.write_int(x)
    elif isinstance(x, float):
        IO.write_float(x)
    elif isinstance(x, str):
        IO.write_string(x)
    elif isinstance(x, list):
        IO.write_list(x)
    elif isinstance(x, ListNode):
        IO.write_linked_list(x)
    elif isinstance(x, TreeNode):
        IO.write_binary_tree(x)
    else:
        print(x, end="")
    print(" ", end="")

{{user template}}

{{Solution Class}}

if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
{{Input Statements}}

File: src/utils/template.java
import java.util.*;
import java.io.*;

// Definition for singly-linked list
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

// Definition for binary tree
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class IO {
    private static Scanner scanner = new Scanner(System.in);
    
    /**
     * Input functions for various data types
     */
    public static boolean readBoolean() {
        String input = scanner.next().toLowerCase();
        return input.equals("true") || input.equals("1");
    }
    
    public static int readInt() {
        return scanner.nextInt();
    }
    
    public static long readLong() {
        return scanner.nextLong();
    }
    
    public static float readFloat() {
        return scanner.nextFloat();
    }
    
    public static double readDouble() {
        return scanner.nextDouble();
    }
    
    public static String readString() {
        return scanner.next();
    }
    
    public static String readLine() {
        scanner.nextLine(); // consume newline
        return scanner.nextLine();
    }
    
    public static int[] readIntArray() {
        String line = scanner.nextLine().trim();
        if (line.equals("[]") || line.isEmpty()) {
            return new int[0];
        }
        
        line = line.substring(1, line.length() - 1); // remove brackets
        if (line.isEmpty()) {
            return new int[0];
        }
        
        String[] parts = line.split(",");
        int[] result = new int[parts.length];
        for (int i = 0; i < parts.length; i++) {
            result[i] = Integer.parseInt(parts[i].trim());
        }
        return result;
    }
    
    public static String[] readStringArray() {
        String line = scanner.nextLine().trim();
        if (line.equals("[]") || line.isEmpty()) {
            return new String[0];
        }
        
        line = line.substring(1, line.length() - 1); // remove brackets
        if (line.isEmpty()) {
            return new String[0];
        }
        
        List<String> result = new ArrayList<>();
        StringBuilder current = new StringBuilder();
        boolean inQuotes = false;
        
        for (char c : line.toCharArray()) {
            if (c == '"' && !inQuotes) {
                inQuotes = true;
            } else if (c == '"' && inQuotes) {
                inQuotes = false;
                result.add(current.toString());
                current = new StringBuilder();
            } else if (c == ',' && !inQuotes) {
                if (current.length() > 0) {
                    result.add(current.toString().trim());
                    current = new StringBuilder();
                }
            } else if (!(c == ' ' && !inQuotes && current.length() == 0)) {
                current.append(c);
            }
        }
        
        if (current.length() > 0) {
            result.add(current.toString().trim());
        }
        
        return result.toArray(new String[0]);
    }
    
    public static int[][] readInt2DArray() {
        String line = scanner.nextLine().trim();
        if (line.equals("[]") || line.isEmpty()) {
            return new int[0][0];
        }
        
        List<int[]> result = new ArrayList<>();
        line = line.substring(1, line.length() - 1); // remove outer brackets
        
        int i = 0;
        while (i < line.length()) {
            if (line.charAt(i) == '[') {
                int j = i + 1;
                int bracketCount = 1;
                while (j < line.length() && bracketCount > 0) {
                    if (line.charAt(j) == '[') {
                        bracketCount++;
                    } else if (line.charAt(j) == ']') {
                        bracketCount--;
                    }
                    j++;
                }
                
                String subArrayStr = line.substring(i + 1, j - 1);
                if (!subArrayStr.isEmpty()) {
                    String[] parts = subArrayStr.split(",");
                    int[] subArray = new int[parts.length];
                    for (int k = 0; k < parts.length; k++) {
                        subArray[k] = Integer.parseInt(parts[k].trim());
                    }
                    result.add(subArray);
                } else {
                    result.add(new int[0]);
                }
                
                i = j;
                while (i < line.length() && (line.charAt(i) == ',' || line.charAt(i) == ' ')) {
                    i++;
                }
            } else {
                i++;
            }
        }
        
        return result.toArray(new int[0][]);
    }
    
    public static ListNode readLinkedList() {
        int[] values = readIntArray();
        if (values.length == 0) {
            return null;
        }
        
        ListNode head = new ListNode(values[0]);
        ListNode current = head;
        for (int i = 1; i < values.length; i++) {
            current.next = new ListNode(values[i]);
            current = current.next;
        }
        return head;
    }
    
    public static TreeNode readBinaryTree() {
        String line = scanner.nextLine().trim();
        if (line.equals("[]") || line.equals("null") || line.isEmpty()) {
            return null;
        }
        
        line = line.substring(1, line.length() - 1); // remove brackets
        if (line.isEmpty()) {
            return null;
        }
        
        String[] values = line.split(",");
        List<Integer> nodeValues = new ArrayList<>();
        
        for (String value : values) {
            String trimmed = value.trim();
            if (trimmed.equals("null")) {
                nodeValues.add(null);
            } else {
                nodeValues.add(Integer.parseInt(trimmed));
            }
        }
        
        if (nodeValues.isEmpty() || nodeValues.get(0) == null) {
            return null;
        }
        
        TreeNode root = new TreeNode(nodeValues.get(0));
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        
        int i = 1;
        while (!queue.isEmpty() && i < nodeValues.size()) {
            TreeNode node = queue.poll();
            
            // Left child
            if (i < nodeValues.size()) {
                if (nodeValues.get(i) != null) {
                    node.left = new TreeNode(nodeValues.get(i));
                    queue.offer(node.left);
                }
                i++;
            }
            
            // Right child
            if (i < nodeValues.size()) {
                if (nodeValues.get(i) != null) {
                    node.right = new TreeNode(nodeValues.get(i));
                    queue.offer(node.right);
                }
                i++;
            }
        }
        
        return root;
    }
    
    /**
     * Output functions for various data types
     */
    public static void writeBoolean(boolean x) {
        System.out.print(x ? "True" : "False");
    }
    
    public static void writeInt(int x) {
        System.out.print(x);
    }
    
    public static void writeLong(long x) {
        System.out.print(x);
    }
    
    public static void writeFloat(float x) {
        System.out.print(x);
    }
    
    public static void writeDouble(double x) {
        System.out.print(x);
    }
    
    public static void writeString(String x) {
        System.out.print(x);
    }
    
    public static void writeIntArray(int[] arr) {
        System.out.print("[");
        for (int i = 0; i < arr.length; i++) {
            if (i > 0) System.out.print(",");
            System.out.print(arr[i]);
        }
        System.out.print("]");
    }
    
    public static void writeStringArray(String[] arr) {
        System.out.print("[");
        for (int i = 0; i < arr.length; i++) {
            if (i > 0) System.out.print(",");
            System.out.print("\"" + arr[i] + "\"");
        }
        System.out.print("]");
    }
    
    public static void writeInt2DArray(int[][] arr) {
        System.out.print("[");
        for (int i = 0; i < arr.length; i++) {
            if (i > 0) System.out.print(",");
            writeIntArray(arr[i]);
        }
        System.out.print("]");
    }
    
    public static void writeLinkedList(ListNode head) {
        List<Integer> values = new ArrayList<>();
        ListNode current = head;
        while (current != null) {
            values.add(current.val);
            current = current.next;
        }
        
        System.out.print("[");
        for (int i = 0; i < values.size(); i++) {
            if (i > 0) System.out.print(",");
            System.out.print(values.get(i));
        }
        System.out.print("]");
    }
    
    public static void writeBinaryTree(TreeNode root) {
        if (root == null) {
            System.out.print("[]");
            return;
        }
        
        List<Integer> result = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        
        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            if (node != null) {
                result.add(node.val);
                queue.offer(node.left);
                queue.offer(node.right);
            } else {
                result.add(null);
            }
        }
        
        // Remove trailing nulls
        while (!result.isEmpty() && result.get(result.size() - 1) == null) {
            result.remove(result.size() - 1);
        }
        
        System.out.print("[");
        for (int i = 0; i < result.size(); i++) {
            if (i > 0) System.out.print(",");
            if (result.get(i) == null) {
                System.out.print("null");
            } else {
                System.out.print(result.get(i));
            }
        }
        System.out.print("]");
    }
    
    // Generic output method
    public static void output(Object obj) {
        if (obj instanceof Boolean) {
            writeBoolean((Boolean) obj);
        } else if (obj instanceof Integer) {
            writeInt((Integer) obj);
        } else if (obj instanceof Long) {
            writeLong((Long) obj);
        } else if (obj instanceof Float) {
            writeFloat((Float) obj);
        } else if (obj instanceof Double) {
            writeDouble((Double) obj);
        } else if (obj instanceof String) {
            writeString((String) obj);
        } else if (obj instanceof int[]) {
            writeIntArray((int[]) obj);
        } else if (obj instanceof String[]) {
            writeStringArray((String[]) obj);
        } else if (obj instanceof int[][]) {
            writeInt2DArray((int[][]) obj);
        } else if (obj instanceof ListNode) {
            writeLinkedList((ListNode) obj);
        } else if (obj instanceof TreeNode) {
            writeBinaryTree((TreeNode) obj);
        } else {
            System.out.print(obj);
        }
        System.out.print(" ");
    }
}

{{user template}}

{{Solution Class}}

public class Main {
    public static void main(String[] args) {
        int t = IO.readInt();
        for (int i = 0; i < t; i++) {
{{Input Statements}}
        }
    }
}